### PR TITLE
Use ML setup from HBW analysis to expand functionality

### PR DESCRIPTION
**Pull request by jomatthi**
This PR expands the machine learning workflow of columnflow by introducing plots during training and more plotting task to study the performance of a trained network. It is taken from the hbw analysis and adapted to run in this analysis.

It is also more convenient to set up different ml models, there's even the basic infrastructure to perform a grid search with different parameters. 

In addition, it was also necessary to fix the way events are categorised due to some earlier columnflow update introducing `categorizer` objects. This has been done already with the "production" categories in PR#9, but needed some further fixes. The new producer `add_ml_cats` has to also be called when producing histograms to make the ml categories available.

An exemplary workflow can be found in `workflow_scripts/dnn_workflow.py`.

### CODE REVIEW

File: mtt/config/categories.py [hunk 1]
```diff
@@ -178,27 +187,50 @@ def add_categories_production(config: od.Config) -> None:
...
+        return
+    logger.debug("Adding ML categories to config using 'add_categories_ml'...")
+
+    # non-ml categories must have been added already
+    if getattr(config.x, "added_categories_production", False) is not True:
+        logger.debug("Production categories not yet present in config, adding them first in 'add_categories_ml'...")
+        add_categories_production(config)
 
     # -- register category selectors
+    from mtt.ml.categories import register_ml_selectors
```
Comment (dsavoiu) [inline mtt/config/categories.py]: Very minor, but why is this moved here rather than importing it at the top? Would do it at the module level unless there is a good reason for having them import in the function itself.

File: mtt/ml/derived/grid_search.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+#     "tt": bkg_weight,
+#     "st": bkg_weight,
+#     "w_lnu": bkg_weight,
+#     "dy": bkg_weight,
+# }
+
+
+example_grid_search = {  # 4*2*2*1*3*3*1 = 144 trainings
+    "layers": [(64, 64), (128, 128), (256, 256), (512, 512)],
+    "learningrate": [0.00500, 0.00050],
```
Comment (dsavoiu) [inline mtt/ml/derived/grid_search.py:30]: We can refine these later of course, but it would be good to make a note here of the (smaller) learning rate we found to converge better in the latest training.

File: mtt/ml/derived/mtt_simple.py [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+
+# how to weight the different classes during training to balance out class imbalances
+class_factors_config = DotDict({
+    "default": DenseClassifier._default__class_factors,
+    "ones": {},  # defaults to 1 (NOTE: do not try to use defaultdict! does not work with hash generation) FIXME?
+    "benchmark_from_hbw": {
+        "tt": 8,
+        "st": 2,
+        "dy": 2,
+        "w_lnu": 1,
```
Comment (dsavoiu) [inline mtt/ml/derived/mtt_simple.py:178]: Conceptual question: what do these factors mean? Does `tt: 8` mean that the network is 8 times more likely to be shown a `tt` event than e.g. `w_lnu`? Or the other way around, i.e. `tt` is 8 times more abundant in the data, so it should be shown 8 times *less* frequently to the network to balance things out?

File: mtt/ml/derived/mtt_simple.py [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+# which configs/eras to use for training
+configs_config = DotDict({
+    "24": lambda self, requested_configs: ["run3_mtt_2024_nano_v15_new"],
+})
+
+# general architecture/ml model parameters
+architecture_config = DotDict({
+    "learning_rate": {
+        "default": DenseClassifier._default__learningrate,
+        "v1_251219": 0.0001,
```
Comment (dsavoiu) [inline mtt/ml/derived/mtt_simple.py]: This is already set in the derivation on L228 -> is it needed here? Better to have a single source of truth for these numbers to avoid confusion later.

File: mtt/ml/helper.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+from columnflow.util import maybe_import
+
+
+ak = maybe_import("awkward")
+np = maybe_import("numpy")
+
+
+def assign_dataset_to_process(
+        dataset_inst: od.Dataset,
+        process_insts: list[od.Process],
```
Comment (dsavoiu) [inline mtt/ml/helper.py]: Minor indentation fix:
```suggestion
    dataset_inst: od.Dataset,
    process_insts: list[od.Process],
```

**Review by dsavoiu — APPROVED** [review]
Hi @jomatthi, thanks a lot for the PR. Looks like a lot of the advanced ML stuff from the HBW is implemented here now as well, which is great! I only have some minor comments (see below), but I'm already approving this so we can merge once they are addressed.

Let's keep the `*_backup` files for now while we finalize the ML setup for `mttbar`, but once that's done I would clean up (they will still be in the Git history if we ever do need them).

File: mtt/config/categories.py (hunk 1 above)
Comment (jomatthi) [reply mtt/config/categories.py]: I don't think there is a good reason. It was probably left over from some debugging step earlier. Will move the import to the top.

File: mtt/ml/derived/grid_search.py (hunk 2 above)
Comment (jomatthi) [reply mtt/ml/derived/grid_search.py:30]: I agree, we could note it somewhere, but I'm not sure this would be a good place. This script just serves as an example on how a gridsearch could be set up. Right now, this file is not included/used, also to avoid instantiating this many ml_model_insts.

File: mtt/ml/derived/mtt_simple.py (hunk 3 above)
Comment (jomatthi) [reply mtt/ml/derived/mtt_simple.py:178]: I'm not quite sure, but if I remember the discussions in previous group meetings correctly, it is the later case. The idea was to show the network less of the more abundant processes and more of the rarer processes. But I'll confirm that with Mathis.

File: mtt/ml/derived/mtt_simple.py (hunk 4 above)
Comment (jomatthi) [reply mtt/ml/derived/mtt_simple.py]: Currently, this dict `architecture_config` is not used at all. It was initially meant to be bookkeeping device as to keep track of parameters that can be changed and were used for specific versions of the network. It is not needed and can be removed to avoid confusion.

**Comment by jomatthi** [comment]
Hi @dsavoiu
Thanks for the review! I've addressed your comments by answering your questions or by committing the suggested changes. I'll go ahead with the merge then.
